### PR TITLE
Restore and fix previous (and more correct) constraint match logic

### LINKED ISSUE #7: Prerelease tags in a version must be rejected if the constraint does not have the exact same semantic version tuple
From https://docs.npmjs.com/misc/semver#prerelease-tags:

> If a version has a prerelease tag (for example, 1.2.3-alpha.3) then it will only be allowed to satisfy comparator sets if at least one comparator with the same [major, minor, patch] tuple also has a prerelease tag.

### PR DESCRIPTION
**Pull request by ixmatus**
Update the unit tests to include tests we expect to fail and tests we expect to succeed when tested for a match.

Some tests have captured possible spec non-conformance when matching a version constraint including a prerelease tag and a version with a prerelease tag that is not equal to but greater than and within the
bounds of the constraint. However, the behavior of the matching logic maintains the important invariants and also maintains the invariant that no version constraint without a prerelease tag matches a version
_with_ a prerelease tag.

Fixes #7. ~but may conflict with (what I think was) the intended fix for #3 and #2.~

The reason why I'm proposing this change is that it _at least_ maintains the invariant that no version with a prerelease tag in it will be matched if a constraint range does not match exactly in the version tuple and have a prerelease tag.

The unit tests, however (see comments in the Units.hs file), reveal some spec non-conformance.

In the case where this makes things difficult with `nixfromnpm`, the new overriding behavior should make it easier to override a package to provide a different version.

This should not be merged until more discussion is had. Notably, I'm also catching more errors in my testing due to this change, I'm not sure yet what it means.

### CODE REVIEW

**Comment by ixmatus** [comment]
Actually, yeah, this is not the correct solution. There are a number of issues to resolve with the prerelease matching logic. I'll see if I can push commits to fix tests that I'm verifying out-of-band with the `semver` CLI utility.

**Comment by adnelson** [comment]
So, pretty much all of those test cases (maybe all?) were copied directly from a list that I had found online. It was a while ago so I'm not sure where I got it from, but some other semver library. The way that prerelease tags work in semver is by far the most confusing part; without them it would be trivial but with them there are many edge cases. Getting this list to pass was my best attempt at covering said cases, but as evidenced by https://github.com/adnelson/nixfromnpm/issues/87, it could be better.

**Comment by ixmatus** [comment]
@adnelson I think I'm pretty close to fixing it. What you had originally was very close. I'm just doing biology experiments on some of the failing tests with the `semver` utility to hone in on what "should" work and what "shouldn't".

**Comment by adnelson** [comment]
Sounds great! However the solution goes, let's make sure that we have the test cases to back it up.

**Comment by ixmatus** [comment]
@adnelson I think I finally nailed it down. The existing test cases pass and I added more test cases.

Please review.

**Comment by ixmatus** [comment]
I can also confirm that this change fixes about 5 or 6 package build errors related to incorrectly resolved versions.

File: src/Data/SemVer/Types.hs [hunk 1]
```diff
@@ -150,6 +150,16 @@ versionsOf = \case
   And svr1 svr2 -> versionsOf svr1 <> versionsOf svr2
   Or svr1 svr2 -> versionsOf svr1 <> versionsOf svr2
 
+stripRangeTags :: SemVerRange -> SemVerRange
```
Comment (adnelson) [inline src/Data/SemVer/Types.hs]: Although it's pretty clear what it's doing, let's drop a doc comment on this guy

File: src/Data/SemVer/Types.hs [hunk 2]
```diff
@@ -195,22 +205,61 @@ renderSV = pack . show
...
+    -- tuple's constraint criteria
+    (Just _, PrereleaseTags []) ->
+      matchesSimple range version
+
+    -- The most important invariant when considering a comparison
+    -- between a range with prerelease tags and a version with
+    -- prerelease tags is whether the semantic version in both is the
+    -- same; if it is not, then we must reject the version.
+    --
+    -- Note that we could have a conjunction, so we want to see if our
```
Comment (adnelson) [inline src/Data/SemVer/Types.hs]: Might be helpful in this comment to show some examples. It's a little difficult to visualize the scenario

File: tests/Unit.hs [hunk 3]
```diff
@@ -178,106 +178,119 @@ rangeTests = describe "range tests" $ do
...
-        ("^1.2.3-alpha", "1.2.3-pre"),
-        ("^1.2.0-alpha", "1.2.0-pre"),
-        ("^0.0.1-alpha", "0.0.1-beta")
+  let testCases :: [(Bool, Text, Text)] = [
+
+        (True, "^1.2.3-alpha.1", "1.2.3-alpha.7"),
+        (True, "^0.0.1-alpha.1", "0.0.1-alpha.t"),
+        (True, "^0.0.1-alpha.1", "0.0.1-alpha.tdff.dddddddddd"),
+        (False, "^1.2.3-alpha", "2.0.0-alpha"),
+        (False, "1.2.3-pre+asdf - 2.4.3-pre+asdf", "1.2.4-pre+asdf"),
```
Comment (adnelson) [inline tests/Unit.hs]: Might be nice given these two `False` examples to show a version that *would* satisfy this range

Comment (adnelson) [reply tests/Unit.hs]: (if there is one)

File: tests/Unit.hs [hunk 4]
```diff
@@ -178,106 +178,119 @@ rangeTests = describe "range tests" $ do
...
-        ("^1.2 ^1", "1.4.2"),
-        ("^1.2.3-alpha", "1.2.3-pre"),
-        ("^1.2.0-alpha", "1.2.0-pre"),
-        ("^0.0.1-alpha", "0.0.1-beta")
+  let testCases :: [(Bool, Text, Text)] = [
+
+        (True, "^1.2.3-alpha.1", "1.2.3-alpha.7"),
+        (True, "^0.0.1-alpha.1", "0.0.1-alpha.t"),
+        (True, "^0.0.1-alpha.1", "0.0.1-alpha.tdff.dddddddddd"),
+        (False, "^1.2.3-alpha", "2.0.0-alpha"),
```
Comment (adnelson) [inline tests/Unit.hs]: The `False` examples involving prerelease tags are the more interesting ones IMO. In the cases that they're testing distinct corner cases, it would be nice to have a comment explaining the particular reason why they should fail

**Review by adnelson — COMMENTED** [review]
A few comment-related tweaks and this is looking good.

File: tests/Unit.hs (hunk 3 above)
Comment (ixmatus) [reply tests/Unit.hs]: There is but it's trivial. I'll add one anyway.

File: tests/Unit.hs (hunk 4 above)
Comment (ixmatus) [reply tests/Unit.hs]: Sure! I'll try to keep the comments brief.

**Comment by ixmatus** [comment]
@adnelson I've addressed your feedback. I also feel good about this being merged now so I removed the `[DO NOT MERGE]` from the title. Though if you have more feedback I'm happy to address it :)

File: tests/Unit.hs [hunk 5]
```diff
@@ -180,21 +180,40 @@ rangeTests = describe "range tests" $ do
...
-        (True, "^1.2.3-alpha.1", "1.2.3-alpha.7"),
-        (True, "^0.0.1-alpha.1", "0.0.1-alpha.t"),
-        (True, "^0.0.1-alpha.1", "0.0.1-alpha.tdff.dddddddddd"),
-        (False, "^1.2.3-alpha", "2.0.0-alpha"),
+        -- Range constraints with pre-release tags require that any
+        -- version satisfying the constraint must be equivalent (in
+        -- its semantic version tuple) to the minimum of all semantic
+        -- versions within the range. In this case the minimum of the
+        -- range is "1.2.3" and the version's semantic version tuple
+        -- is "1.2.3", therefore it does not satisfy the constraints
```
Comment (adnelson) [inline tests/Unit.hs]: Typo: 1.2.4

**Review by adnelson — COMMENTED** [review]
One typo but I’m loving this, I understand it much better myself than I ever did before :)

**Comment by adnelson** [comment]
This is awesome, thanks so much for the hard work!

**Comment by ixmatus** [comment]
Yeah! You're welcome! It was satisfying to solve the problem!
